proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,223 @@
+valid RET, there should be infrastructure to reject this RET and emit a message 
+explaining why it was rejected as an RET.
+
+## Motivation
+Currently, there is an allow list of valid RETs. It must be modified with respect 
+to this spec. Anything that is not a valid RET will be rejected. The allow list isn't
+broad enough, because it doesn't include the case where the RET is user-defined. 
+Ideally, a user should be able to determine how any user-defined structure is invalid 
+as an RET. Some system should be in place to more completely enforce the rules for 
+valid and invalid RETs, as well as provide useful information on why they are invalid.
+
+For example, when targeting DXIL IR, `RWBuffer<double4> b : register(u4);` will emit
+an error in DXC, but will not in clang-dxc, despite the fact that `double4` is an 
+invalid RET for typed buffers.
+
+## Proposed solution
+
+The proposed solution is to use some type_traits defined in the std library, create
+some custom builtins to use as type_traits, and join them together to define a 
+set of conceptual constraints for any RET that is used. These conceptual constraints
+will be applied to every typed buffer resource type that is defined, so that all
+typed buffer HLSL resources have the same rules about which RETs are valid. 
+Validation will occur upon resource type instantiation. Additionally, certain 
+resource types are raw buffer variants, such as `StructuredBuffer`. These resource 
+types will have a different set of type-traits applied, which will loosen constraints
+on viable RETs. The type traits will be validated as `concepts`, which is a feature 
+introduced in C++20 to statically validate a type.
+
+## Detailed design
+
+In `clang\lib\Sema\HLSLExternalSemaSource.cpp`, `RWBuffer` is defined, along with 
+`RasterizerOrderedBuffer` and `StructuredBuffer`. It is at this point that the 
+`type_traits` should be incorporated into these resource declarations. A concept 
+containing the relevant `type_traits` will be applied to each resource declaration,
+and there wil be sufficient context to determine the target IR.
+If DXIL is given, then all of the typed buffer `type_traits` will be applied on each
+typed buffer HLSL resource type. Otherwise, the raw buffer type_traits will be 
+applied to each resource type. If a `type_trait` is not true for the given 
+RET, a corresponding error message will be emitted.
+
+The list of type_traits that will be available for use are described below:
+| type_trait | Description|
+|-|-|
+| `__is_complete_type` | An RET should either be a complete type, or a user defined type that has been completely defined. |
+| `__is_intangible` | An RET should be an arithmetic type, or a bool, or a vector or matrix or UDT containing such types. This is equivalent to validating that the RET is not intangible. |
+| `__builtin_is_homogenous` | A typed buffer RET with the DXIL IR target should never have two different subelement types. |

```

Comment:
```__builtin_is_homogenous``` will have an interesting definition. For example, for it to be useful here, it'll need to return true here:

```
struct Foo {
    int2 x;
};

struct Type {
    Foo f;
    int2 x;
};
```

However, with a very generalized name like this I think that might be confusing for other scenarios.

Reply:
Yes, its definition will use a function that's already been defined, that collects all the leaf types and puts them in a small vector.  The first element in this small vector will be compared to each subsequent element, and any mismatches will result in false, true otherwise. 
Other names that could be considered: __builtin_is_leaf_type_homogenous, __builtin_is_sub_element_homogenous?